Python/Product/TestAdapter.Executor/UnitTest/TestExecutorUnitTest.cs

```diff
@@ -83,64 +85,32 @@ class TestExecutorUnitTest : ITestExecutor {
-            var executorUri = new Uri(PythonConstants.TestExecutorUriString);
-            var tests = new List<TestCase>();
-            var doc = Read(runContext.RunSettings.SettingsXml);
-            foreach (var t in TestReader.ReadTests(doc, new HashSet<string>(sources, StringComparer.OrdinalIgnoreCase), m => {
-                frameworkHandle?.SendMessage(TestMessageLevel.Warning, m);
-            })) {
-                tests.Add(new TestCase(t.FullyQualifiedName, executorUri, t.SourceFile) {
-                    DisplayName = t.DisplayName,
-                    LineNumber = t.LineNo,
-                    CodeFilePath = t.SourceFile
-                });
-            }
+            var sourceToProjSettings = RunSettingsUtil.GetSourceToProjSettings(runContext.RunSettings);
 
-            if (_cancelRequested.WaitOne(0)) {
-                return;
-            }
+            var testColletion = new TestCollection();
 
-            RunTestCases(tests, runContext, frameworkHandle);
-        }
+            foreach (var testGroup in sources.GroupBy(x => sourceToProjSettings[x])) {
+                var settings = testGroup.Key;
 
-        private Dictionary<string, PythonProjectSettings> GetSourceToSettings(IRunSettings settings) {
-            var doc = Read(settings.SettingsXml);
-            XPathNodeIterator nodes = doc.CreateNavigator().Select("/RunSettings/Python/TestCases/Project");
-            Dictionary<string, PythonProjectSettings> res = new Dictionary<string, PythonProjectSettings>();
-
-            foreach (XPathNavigator project in nodes) {
-                PythonProjectSettings projSettings = new PythonProjectSettings(
-                    project.GetAttribute("home", ""),
-                    project.GetAttribute("workingDir", ""),
-                    project.GetAttribute("interpreter", ""),
-                    project.GetAttribute("pathEnv", ""),
-                    project.GetAttribute("nativeDebugging", "").IsTrue(),
-                    project.GetAttribute("useLegacyDebugger", "").IsTrue()
-                );
-
-                foreach (XPathNavigator environment in project.Select("Environment/Variable")) {
-                    projSettings.Environment[environment.GetAttribute("name", "")] = environment.GetAttribute("value", "");
-                }
-
-                string djangoSettings = project.GetAttribute("djangoSettingsModule", "");
-                if (!String.IsNullOrWhiteSpace(djangoSettings)) {
-                    projSettings.Environment["DJANGO_SETTINGS_MODULE"] = djangoSettings;
-                }
-
-                foreach (XPathNavigator searchPath in project.Select("SearchPaths/Search")) {
-                    projSettings.SearchPath.Add(searchPath.GetAttribute("value", ""));
+                try {
+                    var discovery = DiscovererFactory.GetDiscoverer(settings);
+                    discovery.DiscoverTests(testGroup, frameworkHandle, testColletion);
+                } catch (Exception ex) {
```

Comment:
when vstest execution is called from commandline we only get source files and we need to do discovery first and store tests in a testCollection.